File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/view/main/ExpensesRecyclerViewAdapter.java
/*
 *   Copyright 2015 Benoit LETONDOR
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.benoitletondor.easybudgetapp.view.main;

import android.app.Activity;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.benoitletondor.easybudgetapp.R;
import com.benoitletondor.easybudgetapp.helper.CurrencyHelper;
import com.benoitletondor.easybudgetapp.helper.Parameters;
import com.benoitletondor.easybudgetapp.model.Expense;
import com.benoitletondor.easybudgetapp.model.RecurringExpenseDeleteType;
import com.benoitletondor.easybudgetapp.view.expenseedit.ExpenseEditActivity;

import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

/**
 * Recycler view adapter to display expenses for a given date
 *
 * @author Benoit LETONDOR
 */
public class ExpensesRecyclerViewAdapter extends RecyclerView.Adapter<ExpensesRecyclerViewAdapter.ViewHolder>
{
    @NonNull
    private List<Expense> expenses = new ArrayList<>(0);
    @NonNull
    private Date date;
    @NonNull
    private final Activity activity;
    @NonNull
    private final Parameters parameters;

// ------------------------------------------->

    /**
     * Instanciate an adapter for the given date
     */
    public ExpensesRecyclerViewAdapter(@NonNull Activity activity, @NonNull Parameters parameters, @NonNull Date date)
    {
        this.activity = activity;
        this.parameters = parameters;
        this.date = date;
    }

    /**
     * Return the date content is displayed for displayed
     */
    @NonNull
    public Date getDate()
    {
        return date;
    }

    /**
     * Set a new date and data to display
     */
    public void setDate(@NonNull Date date, @NonNull List<Expense> expenses)
    {
        this.date = date;
        this.expenses = expenses;
        notifyDataSetChanged();
    }

    /**
     * Remove given expense
     *
     * @param expense the expense to remove
     * @return position of the deleted expense (-1 if not found)
     */
    public int removeExpense(Expense expense)
    {
        Iterator<Expense> expenseIterator = expenses.iterator();
        int position = 0;
        while( expenseIterator.hasNext() )
        {
            Expense shownExpense = expenseIterator.next();
            if( shownExpense.getId().equals(expense.getId()) )
            {
                expenseIterator.remove();
                notifyItemRemoved(position);
                return position;
            }

            position++;
        }

        return -1;
    }

    /**
     * Add an expense at the given position
     */
    public void addExpense(Expense expense, int position)
    {
        expenses.add(position, expense);
        notifyItemRangeInserted(position, 1);
    }

// ------------------------------------------>

    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i)
    {
        View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.recycleview_expense_cell, viewGroup, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder viewHolder, int i)
    {
        final Expense expense = expenses.get(i);

        viewHolder.expenseTitleTextView.setText(expense.getTitle());
        viewHolder.expenseAmountTextView.setText(CurrencyHelper.getFormattedCurrencyString(parameters, -expense.getAmount()));
        viewHolder.expenseAmountTextView.setTextColor(ContextCompat.getColor(viewHolder.view.getContext(), expense.isRevenue() ? R.color.budget_green : R.color.budget_red));
        viewHolder.recurringIndicator.setVisibility(expense.isRecurring() ? View.VISIBLE : View.GONE);
        viewHolder.positiveIndicator.setImageResource(expense.isRevenue() ? R.drawable.ic_label_green : R.drawable.ic_label_red);

        if( expense.isRecurring() )
        {
            assert expense.getAssociatedRecurringExpense() != null;
            switch (expense.getAssociatedRecurringExpense().getType())
            {
                case WEEKLY:
                    viewHolder.recurringIndicatorTextview.setText(viewHolder.view.getContext().getString(R.string.weekly));
                    break;
                case BI_WEEKLY:
                    viewHolder.recurringIndicatorTextview.setText(viewHolder.view.getContext().getString(R.string.bi_weekly));
                    break;
                case MONTHLY:
                    viewHolder.recurringIndicatorTextview.setText(viewHolder.view.getContext().getString(R.string.monthly));
                    break;
                case YEARLY:
                    viewHolder.recurringIndicatorTextview.setText(viewHolder.view.getContext().getString(R.string.yearly));
                    break;
            }
        }

        final View.OnClickListener onClickListener = v -> {
            if (expense.isRecurring())
            {
                AlertDialog.Builder builder = new AlertDialog.Builder(activity);
                builder.setTitle(expense.isRevenue() ? R.string.dialog_edit_recurring_income_title : R.string.dialog_edit_recurring_expense_title);
                builder.setItems(expense.isRevenue() ? R.array.dialog_edit_recurring_income_choices : R.array.dialog_edit_recurring_expense_choices, (dialog, which) -> {
                    switch ( which )
                    {
                        case 0: // Edit this one
                        {
                            Intent startIntent = new Intent(viewHolder.view.getContext(), ExpenseEditActivity.class);
                            startIntent.putExtra("date", expense.getDate().getTime());
                            startIntent.putExtra("expense", expense);

                            ActivityCompat.startActivityForResult(activity, startIntent, MainActivity.ADD_EXPENSE_ACTIVITY_CODE, null);

                            break;
                        }
                        case 1: // Delete this one
                        {
                            // Send notification to inform views that this expense has been deleted
                            Intent intent = new Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED);
                            intent.putExtra("expense", expense);
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.ONE.getValue());
                            LocalBroadcastManager.getInstance(activity.getApplicationContext()).sendBroadcast(intent);

                            break;
                        }
                        case 2: // Delete from
                        {
                            // Send notification to inform views that this expense has been deleted
                            Intent intent = new Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED);
                            intent.putExtra("expense", expense);
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.FROM.getValue());
                            LocalBroadcastManager.getInstance(activity.getApplicationContext()).sendBroadcast(intent);

                            break;
                        }
                        case 3: // Delete up to
                        {
                            // Send notification to inform views that this expense has been deleted
                            Intent intent = new Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED);
                            intent.putExtra("expense", expense);
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.TO.getValue());
                            LocalBroadcastManager.getInstance(activity.getApplicationContext()).sendBroadcast(intent);

                            break;
                        }
                        case 4: // Delete all
                        {
                            // Send notification to inform views that this expense has been deleted
                            Intent intent = new Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED);
                            intent.putExtra("expense", expense);
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.ALL.getValue());
                            LocalBroadcastManager.getInstance(activity.getApplicationContext()).sendBroadcast(intent);

                            break;
                        }
                    }
                });
                builder.show();
            }
            else
            {
                AlertDialog.Builder builder = new AlertDialog.Builder(activity);
                builder.setTitle(expense.isRevenue() ? R.string.dialog_edit_income_title : R.string.dialog_edit_expense_title);
                builder.setItems(expense.isRevenue() ? R.array.dialog_edit_income_choices : R.array.dialog_edit_expense_choices, (dialog, which) -> {
                    switch ( which )
                    {
                        case 0: // Edit expense
                        {
                            Intent startIntent = new Intent(viewHolder.view.getContext(), ExpenseEditActivity.class);
                            startIntent.putExtra("date", expense.getDate().getTime());
                            startIntent.putExtra("expense", expense);

                            ActivityCompat.startActivityForResult(activity, startIntent, MainActivity.ADD_EXPENSE_ACTIVITY_CODE, null);

                            break;
                        }
                        case 1: // Delete
                        {
                            // Send notification to inform views that this expense has been deleted
                            Intent intent = new Intent(MainActivity.INTENT_EXPENSE_DELETED);
                            intent.putExtra("expense", expense);
                            LocalBroadcastManager.getInstance(activity.getApplicationContext()).sendBroadcast(intent);

                            break;
                        }
                    }
                });
                builder.show();
            }

        };

        viewHolder.view.setOnClickListener(onClickListener);

        viewHolder.view.setOnLongClickListener(v -> {
            onClickListener.onClick(v);
            return true;
        });
    }

    @Override
    public int getItemCount()
    {
        return expenses.size();
    }

// ------------------------------------------->

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    static class ViewHolder extends RecyclerView.ViewHolder
    {
        final TextView expenseTitleTextView;
        final TextView expenseAmountTextView;
        final ViewGroup recurringIndicator;
        final TextView recurringIndicatorTextview;
        final ImageView positiveIndicator;
        final View view;

        ViewHolder(View v)
        {
            super(v);

            view = v;
            expenseTitleTextView = v.findViewById(R.id.expense_title);
            expenseAmountTextView = v.findViewById(R.id.expense_amount);
            recurringIndicator = v.findViewById(R.id.recurring_indicator);
            recurringIndicatorTextview = v.findViewById(R.id.recurring_indicator_textview);
            positiveIndicator = v.findViewById(R.id.positive_indicator);
        }
    }
}

File: Android/EasyBudget/app/src/main/java/com/benoitletondor/easybudgetapp/view/main/ExpensesRecyclerViewAdapter.kt
/*
 *   Copyright 2019 Benoit LETONDOR
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package com.benoitletondor.easybudgetapp.view.main

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.RecyclerView
import com.benoitletondor.easybudgetapp.R
import com.benoitletondor.easybudgetapp.helper.CurrencyHelper
import com.benoitletondor.easybudgetapp.model.Expense
import com.benoitletondor.easybudgetapp.model.RecurringExpenseDeleteType
import com.benoitletondor.easybudgetapp.model.RecurringExpenseType
import com.benoitletondor.easybudgetapp.parameters.Parameters
import com.benoitletondor.easybudgetapp.view.expenseedit.ExpenseEditActivity
import java.util.*

/**
 * Recycler view adapter to display expenses for a given date
 *
 * @author Benoit LETONDOR
 */
class ExpensesRecyclerViewAdapter(private val activity: Activity,
                                  private val parameters: Parameters,
                                  private var date: Date) : RecyclerView.Adapter<ExpensesRecyclerViewAdapter.ViewHolder>() {

    private var expenses = mutableListOf<Expense>()

    fun getDate(): Date = date

    /**
     * Set a new date and data to display
     */
    fun setDate(date: Date, expenses: List<Expense>) {
        this.date = date
        this.expenses.clear()
        this.expenses.addAll(expenses)
        notifyDataSetChanged()
    }

    /**
     * Remove given expense
     *
     * @param expense the expense to remove
     * @return position of the deleted expense (-1 if not found)
     */
    fun removeExpense(expense: Expense): Int {
        val expenseIterator = expenses.iterator()
        var position = 0
        while (expenseIterator.hasNext()) {
            val (id) = expenseIterator.next()
            if (expense.id != null && expense.id == id) {
                expenseIterator.remove()
                notifyItemRemoved(position)
                return position
            }

            position++
        }

        return -1
    }

// ------------------------------------------>

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.recycleview_expense_cell, viewGroup, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        val expense = expenses[i]

        viewHolder.expenseTitleTextView.text = expense.title
        viewHolder.expenseAmountTextView.text = CurrencyHelper.getFormattedCurrencyString(parameters, -expense.amount)
        viewHolder.expenseAmountTextView.setTextColor(ContextCompat.getColor(viewHolder.view.context, if (expense.isRevenue()) R.color.budget_green else R.color.budget_red))
        viewHolder.recurringIndicator.visibility = if (expense.isRecurring()) View.VISIBLE else View.GONE
        viewHolder.positiveIndicator.setImageResource(if (expense.isRevenue()) R.drawable.ic_label_green else R.drawable.ic_label_red)

        if (expense.isRecurring()) {
            when (expense.associatedRecurringExpense!!.type) {
                RecurringExpenseType.DAILY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.daily)
                RecurringExpenseType.WEEKLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.weekly)
                RecurringExpenseType.BI_WEEKLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.bi_weekly)
                RecurringExpenseType.TER_WEEKLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.ter_weekly)
                RecurringExpenseType.FOUR_WEEKLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.four_weekly)
                RecurringExpenseType.MONTHLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.monthly)
                RecurringExpenseType.YEARLY -> viewHolder.recurringIndicatorTextview.text = viewHolder.view.context.getString(R.string.yearly)
            }
        }

        val onClickListener = View.OnClickListener {
            if (expense.isRecurring()) {
                val builder = AlertDialog.Builder(activity)
                builder.setTitle(if (expense.isRevenue()) R.string.dialog_edit_recurring_income_title else R.string.dialog_edit_recurring_expense_title)
                builder.setItems(if (expense.isRevenue()) R.array.dialog_edit_recurring_income_choices else R.array.dialog_edit_recurring_expense_choices) { _, which ->
                    when (which) {
                        // Edit this one
                        0 -> {
                            val startIntent = Intent(viewHolder.view.context, ExpenseEditActivity::class.java)
                            startIntent.putExtra("date", expense.date.time)
                            startIntent.putExtra("expense", expense)

                            ActivityCompat.startActivityForResult(activity, startIntent, MainActivity.ADD_EXPENSE_ACTIVITY_CODE, null)
                        }
                        // Delete this one
                        1 -> {
                            // Send notification to inform views that this expense has been deleted
                            val intent = Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED)
                            intent.putExtra("expense", expense)
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.ONE.value)
                            LocalBroadcastManager.getInstance(activity.applicationContext).sendBroadcast(intent)
                        }
                        // Delete from
                        2 -> {
                            // Send notification to inform views that this expense has been deleted
                            val intent = Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED)
                            intent.putExtra("expense", expense)
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.FROM.value)
                            LocalBroadcastManager.getInstance(activity.applicationContext).sendBroadcast(intent)
                        }
                        // Delete up to
                        3 -> {
                            // Send notification to inform views that this expense has been deleted
                            val intent = Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED)
                            intent.putExtra("expense", expense)
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.TO.value)
                            LocalBroadcastManager.getInstance(activity.applicationContext).sendBroadcast(intent)
                        }
                        // Delete all
                        4 -> {
                            // Send notification to inform views that this expense has been deleted
                            val intent = Intent(MainActivity.INTENT_RECURRING_EXPENSE_DELETED)
                            intent.putExtra("expense", expense)
                            intent.putExtra("deleteType", RecurringExpenseDeleteType.ALL.value)
                            LocalBroadcastManager.getInstance(activity.applicationContext).sendBroadcast(intent)
                        }
                    }
                }
                builder.show()
            } else {
                val builder = AlertDialog.Builder(activity)
                builder.setTitle(if (expense.isRevenue()) R.string.dialog_edit_income_title else R.string.dialog_edit_expense_title)
                builder.setItems(if (expense.isRevenue()) R.array.dialog_edit_income_choices else R.array.dialog_edit_expense_choices) { _, which ->
                    when (which) {
                        0 // Edit expense
                        -> {
                            val startIntent = Intent(viewHolder.view.context, ExpenseEditActivity::class.java)
                            startIntent.putExtra("date", expense.date.time)
                            startIntent.putExtra("expense", expense)

                            ActivityCompat.startActivityForResult(activity, startIntent, MainActivity.ADD_EXPENSE_ACTIVITY_CODE, null)
                        }
                        1 // Delete
                        -> {
                            // Send notification to inform views that this expense has been deleted
                            val intent = Intent(MainActivity.INTENT_EXPENSE_DELETED)
                            intent.putExtra("expense", expense)
                            LocalBroadcastManager.getInstance(activity.applicationContext).sendBroadcast(intent)
                        }
                    }
                }
                builder.show()
            }

        }

        viewHolder.view.setOnClickListener(onClickListener)

        viewHolder.view.setOnLongClickListener { v ->
            onClickListener.onClick(v)
            true
        }
    }

    override fun getItemCount(): Int = expenses.size

// ------------------------------------------->

    class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
        val expenseTitleTextView: TextView = view.findViewById(R.id.expense_title)
        val expenseAmountTextView: TextView = view.findViewById(R.id.expense_amount)
        val recurringIndicator: ViewGroup = view.findViewById(R.id.recurring_indicator)
        val recurringIndicatorTextview: TextView = view.findViewById(R.id.recurring_indicator_textview)
        val positiveIndicator: ImageView = view.findViewById(R.id.positive_indicator)
    }
}